lib/ui/window.dart

```diff
@@ -365,11 +407,95 @@ class Window {
   /// Message handlers must call the function given in the `callback` parameter.
   /// If the handler does not need to respond, the handler should pass `null` to
   /// the callback.
-  PlatformMessageCallback onPlatformMessage;
+  PlatformMessageCallback get onPlatformMessage => _onPlatformMessage;
+
+  set onPlatformMessage(PlatformMessageCallback callback) {
+    _onPlatformMessage = _zonedPlatformMessageCallback(callback);
+  }
+
+  PlatformMessageCallback _onPlatformMessage;
 
   /// Called by [_dispatchPlatformMessage].
   void _respondToPlatformMessage(int responseId, ByteData data)
       native "Window_respondToPlatformMessage";
+
+  /// Wraps the given [callback] in another callback that ensures that the
+  /// original callback is called in the zone it was registered in.
+  static VoidCallback _zonedVoidCallback(VoidCallback callback) {
```

Comment:
The problem with this approach is that it doesn't maintain this invariant:
```dart
  foo = bar;
  assert(foo == bar);
```
...which I think we should maintain. I don't have an immediate solution for this unfortunately. Is there no way to introspect the callbacks anywhere else that uses zones?